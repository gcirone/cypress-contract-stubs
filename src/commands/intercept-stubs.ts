import type { RouteMatcherOptions, CyHttpMessages } from 'cypress/types/net-stubbing';
import type { StubEntries, StubEntry, StubRequest } from '../plugin/stubs/stubs-entries';
import jsonpath from 'jsonpath';

declare global {
  namespace Cypress {
    interface InterceptStubsOptions {
      names?: string[];
      headers?: Record<string, any>;
      matchRequestHandler?: (request: CyHttpMessages.IncomingRequest) => boolean;
    }

    interface Chainable {
      interceptStubs(options?: InterceptStubsOptions): Chainable<null>;
    }
  }
}

export function interceptStubs(options?: Cypress.InterceptStubsOptions) {
  cy.task<StubEntries>('contract:stubs', null, { log: false }).then((stubs) => {
    stubs
      .filter((stub) => filterByStubNames(stub.name, options?.names))
      .filter((stub) => filterByStubOptions(stub, options))
      .sort(sortByPriority)
      .forEach(({ id, name, request, response }) => {
        try {
          const matcher: RouteMatcherOptions = { query: {}, headers: {} };
          matcher.method = request.method;
          matchRequestUrl(request, matcher);
          matchRequestQueryParams(request, matcher);
          matchRequestHeaders(request, matcher);

          cy.intercept(matcher, (req) => {
            const replayWithStubResponse = () => {
              const { status, body, headers } = response;
              req.reply(status || 200, body, headers);
            };

            if (options?.matchRequestHandler) {
              if (options.matchRequestHandler(req)) {
                replayWithStubResponse();
              }
            } else {
              if (request.bodyPatterns) {
                if (matchRequestBody(request, JSON.parse(req.body))) {
                  replayWithStubResponse();
                }
              } else {
                replayWithStubResponse();
              }
            }
          }).as(id);
        } catch (e: any) {
          console.error(`Error when generating matcher for stub "${name}"`, e.message);
        }
      });
  });
}

function filterByStubNames(name: string, stubNames?: string[]) {
  if (stubNames?.length) {
    return stubNames.includes(name);
  } else {
    return true;
  }
}

function filterByStubOptions({ request }: StubEntry, options?: Cypress.InterceptStubsOptions) {
  let shouldBeIntercepted = true;

  for (const header in options?.headers) {
    const optionsValue = options?.headers[header];
    const stubValue = request.headers?.[header];

    if (stubValue?.equalTo) {
      if (stubValue.equalTo === optionsValue) {
        delete request.headers?.[header];
        shouldBeIntercepted = true;
      } else {
        shouldBeIntercepted = false;
      }
    } else if (stubValue?.matches) {
      if (matchPattern(stubValue.matches).test(optionsValue)) {
        delete request.headers?.[header];
        shouldBeIntercepted = true;
      } else {
        shouldBeIntercepted = false;
      }
    }
  }

  return shouldBeIntercepted;
}

function sortByPriority(stubA: StubEntry, stubB: StubEntry): number {
  const priorityA = stubA.priority || 0;
  const priorityB = stubB.priority || 0;
  return priorityA - priorityB;
}

function matchRequestUrl(request: StubRequest, matcher: RouteMatcherOptions) {
  if (request.url) {
    matcher.url = request.url;
  } else if (request.urlPath) {
    matcher.path = request.urlPath;
  } else if (request.urlPattern) {
    matcher.url = matchPattern(request.urlPattern);
  } else if (request.urlPathPattern) {
    matcher.path = matchPattern(request.urlPathPattern);
  }
}

function matchRequestQueryParams(request: StubRequest, matcher: RouteMatcherOptions) {
  for (const param in request.queryParameters) {
    const value = request.queryParameters[param];

    if (value.equalTo) {
      matcher.query![param] = value.equalTo;
    } else if (value.matches) {
      matcher.query![param] = matchPattern(value.matches);
    }
  }
}

function matchRequestHeaders(request: StubRequest, matcher: RouteMatcherOptions) {
  for (const header in request.headers) {
    const value = request.headers[header];

    if (value.equalTo) {
      matcher.headers![header] = value.equalTo;
    } else if (value.matches) {
      matcher.headers![header] = matchPattern(value.matches);
    }
  }
}

function matchRequestBody(request: StubRequest, body: any) {
  const shouldReplyWithStub: boolean[] = [];
  
  for (const { matchesJsonPath } of request?.bodyPatterns || []) {
    const { expression, equalTo, equalToJson } = matchesJsonPath;
    const [value] = expression && jsonpath.query(body, expression);

    if (equalTo && value) {
      shouldReplyWithStub.push(equalTo === value);
    } else if (equalToJson && value) {
      shouldReplyWithStub.push(equalToJson === JSON.stringify(value));
    }
  }

  return shouldReplyWithStub.every((r) => r);
}

function matchPattern(pattern: string) {
  return new RegExp(pattern.replace('?+', '?.*'), 'g');
}

Cypress.Commands.add('interceptStubs', interceptStubs);
