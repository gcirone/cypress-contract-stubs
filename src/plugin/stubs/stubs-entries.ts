export interface ArchiveEntry {
  path: string;
  type: number;
}

export interface StubMatcher {
  equalTo?: string;
  equalToJson?: string;
  matches?: string;
  expression?: string;
}

export interface StubRequest {
  method?: string;
  url?: string;
  urlPath?: string;
  urlPattern?: string;
  urlPathPattern?: string;
  queryParameters?: Record<string, StubMatcher>;
  headers?: Record<string, StubMatcher>;
  bodyPatterns?: { matchesJsonPath: StubMatcher }[];
}

export interface StubResponse {
  status?: number;
  body?: any;
  headers?: Record<string, string>;
  transformers?: string[];
}

export interface StubEntry {
  id: string;
  request: StubRequest;
  response: StubResponse;
  uuid: string;
  name: string;
  consumer?: string;
  priority?: number;
}

export type StubEntries = StubEntry[];

/**
 * Array containing all stubs entries
 */
export const stubEntries: StubEntries = [];

/**
 * Append stub entries
 *
 * @param stubs
 */
export function storeStubEntries(stubs: StubEntries): void {
  stubEntries.push(...stubs);
}
