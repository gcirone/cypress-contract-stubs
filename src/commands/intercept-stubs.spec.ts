import { interceptStubs } from './intercept-stubs';
import { mockStubs } from './__mocks__/stubs.mocks';

describe('Intercept Stubs Command', () => {
  const reqReplySpy = jest.fn();
  const asSpy = jest.fn();
  let mockedBody: any;

  beforeEach(() => {
    jest.spyOn(cy, 'task').mockResolvedValueOnce(mockStubs);
    jest.spyOn<any, any>(cy, 'intercept').mockImplementation((_, response: any) => {
      response({ body: mockedBody, reply: reqReplySpy });
      return { as: asSpy };
    });
  });

  afterEach(() => jest.clearAllMocks());

  it('should register the cypress commands', () => {
    expect(Cypress.Commands.add).toHaveBeenCalledWith('interceptStubs', interceptStubs);
  });

  it('#interceptStubs should intercept all request with stubs', async () => {
    await interceptStubs();

    expect(cy.task).toHaveBeenCalledWith('contract:stubs', null, { log: false });

    expect(cy.intercept).toHaveBeenCalledTimes(5);
    expect((cy.intercept as jest.Mock).mock.calls[2][0]).toMatchSnapshot();

    expect(reqReplySpy).toHaveBeenCalledTimes(5);
    expect(reqReplySpy).toHaveBeenNthCalledWith(3, 204, 'c', { hr: 'x' });

    expect(asSpy).toHaveBeenCalledTimes(5);
  });

  it('#interceptStubs should intercept only the named stubs', async () => {
    await interceptStubs({ names: ['b'] });

    expect(cy.task).toHaveBeenCalledWith('contract:stubs', null, { log: false });
    expect((cy.intercept as jest.Mock).mock.calls[0][0]).toMatchSnapshot();

    expect(asSpy).toHaveBeenCalledWith('b:f119');
  });

  it('#interceptStubs should intercept only the named stubs with options', async () => {
    await interceptStubs({ names: ['b'], headers: { tpl: 'tpl2' } });

    expect(cy.intercept).toHaveBeenCalledTimes(2);
    expect(asSpy).toHaveBeenCalledTimes(2);
    expect(asSpy).toHaveBeenNthCalledWith(1, 'b:f119');
    expect(asSpy).toHaveBeenNthCalledWith(2, 'b:5163');
  });

  it.only('#interceptStubs should intercept named stubs with body', async () => {
    mockedBody = JSON.stringify({ operationName: 'GetPost', variables: { id: 1 } });
    await interceptStubs({ names: ['d'] });
  });
});
