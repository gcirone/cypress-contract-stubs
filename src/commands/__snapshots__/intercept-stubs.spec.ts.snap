// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Intercept Stubs Command #interceptStubs should intercept all request with stubs 1`] = `
Object {
  "headers": Object {
    "h1": /\\.\\*/g,
    "h2": "2",
  },
  "method": "POST",
  "pathname": /\\\\/path\\\\/\\?\\.\\*other/g,
  "query": Object {
    "q1": "1",
    "q2": /\\.\\*/g,
  },
}
`;

exports[`Intercept Stubs Command #interceptStubs should intercept only the named stubs 1`] = `
Object {
  "headers": Object {},
  "method": "GET",
  "query": Object {
    "q1": "1",
  },
  "url": /\\.\\*/g,
}
`;
